Auto-linking for native modules (source-based)

jonthysell opened the pull request:
* Updated `npx react-native config` to create the needed json for auto-linking
* New `npx react-native autolink-windows` command to perform auto-linking
* Auto-linking occurs automatically during `npx react-native run-windows`
* New Auto-linking targets and props to catch so RNW app builds (in VS or with `run-windows --no-autolink`) throw a warning that auto-linking should be run
* Update external RNW lib targets and props so `run-windows` works
* Fixed `run-windows` to properly consume react-native config
* Updated in-repo apps for autolinking:
  * E2ETest, SampleAppCpp, and SampleAppCS will register autolinked modules
  * playground has the generated files in the project, but does not use them
  * playground-win32 has the generated files but does not include them in the project

Closes #2853

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/5044)

asklar commented on vnext/PropertySheets/Autolink.targets:
```diff
@@ -0,0 +1,10 @@
+-->
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Target Name="RunAutolink" BeforeTargets="PrepareForBuild">
+    <Exec Command="$(AutolinkCommand) $(AutolinkCommandArgs)" ConsoleToMSBuild="true" WorkingDirectory="$(AutolinkCommandWorkingDir)" CustomWarningRegularExpression="Warning: " CustomErrorRegularExpression="Error: " />
```
`ConsoleToMSBuild` set to true will make the output available into whatever the `ConsoleOutput` attribute gets set to, so should we capture it by setting `ConsoleOutput` to something and then Issuing a <Error /> element with that variable as text if we find an error?

jonthysell replied:
I just copied this from the BundleCommand in the Bundle.Common.targets. I just want to make sure that the warning/errors get parsed and displayed in VS.

jonthysell replied:
Removed, still does what I need.

asklar commented on vnext/PropertySheets/External/Microsoft.ReactNative.Uwp.CSharpModule.props:
```diff
@@ -0,0 +1,10 @@
+-->
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <ReactNativeWindowsDir Condition="'$(ReactNativeWindowsDir)' == ''">$([MSBuild]::NormalizePath('$(MSBuildThisFileDirectory)\..\..'))\</ReactNativeWindowsDir>
```
can you clarify where this props goes? is this going into node-modules? isn't this missing react-native-windows at the end?

jonthysell replied:
I plan on adding new module templates to the CLI, and just like we just added targets/props for new cli apps to consume, these props/targets are for new cli modules. I also plan on updating existing community modules to load these targets, which I've done locally for react-native-camera.

jonthysell replied:
Someone added Libs targets, switching to them.

asklar commented on vnext/PropertySheets/External/Microsoft.ReactNative.Uwp.CSharpModule.targets:
```diff
@@ -0,0 +1,21 @@
+      <Project>{F2824844-CE15-4242-9420-308923CD76C3}</Project>
+      <Name>Microsoft.ReactNative.Managed</Name>
+    </ProjectReference>
+    <PackageReference Include="Microsoft.UI.Xaml">
```
is this needed?

jonthysell replied:
The `Microsoft.ReactNative.Manage` reference, yes. That's where all of the native module reflection (and future code-gen) is.

The `Microsoft.UI.Xaml` reference, I'm not 100% sure. If you use xaml in your native module lib, and the app uses mux, is it okay if the lib doesn't use mux?

asklar replied:
yeah meant the MUX reference (I might need a "this is how to use GH for PRs" tutorial 😿).
I think this should be fine to remove. Not every module will need MUX (esp if they don't use any controls in MUX but just use the built-in WUX stuff)

jonthysell replied:
So how do we make sure they don't use a different version of mux than we do? We enforce it for the app right?

asklar replied:
We shouldn't require any MUX controls in the core. We will for WinUI3 but for everyone on system XAML, MUX should be not needed. Even if we were to have a MUX 2.x dependency from core, it _should_ be ok for the app to have its own version (they'll just be bloaty, hopefully realize that they're loading code "2x" and fix the import in their core or in the app)

jonthysell replied:
Fixed.

asklar commented on vnext/local-cli/config/configUtils.js:
```diff
@@ -0,0 +1,144 @@
+    const solutionName = packageJson.name;
+
+    // First look for a solution named solutionName.sln
+    const slnFile = findFiles(winFolder, solutionName + '.sln')[0];
```
actually we used to do this in the CLI and just choose the first one, and that was causing issues because playground has both a win32 and uwp SLN in the same folder. I updated the code to bail out if there is more than one SLN since it's not predictable what solution you'd end up building. This can be overridden with the `--sln`  command line argument.
Let's consolidate this code and that to use the same code?

jonthysell replied:
Fixed.

asklar commented on vnext/local-cli/config/configUtils.js:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  // Next look for any project *.vcxproj
+  const allCppProj = findFiles(winFolder, '*.vcxproj');
```
I'm not a fan of the "choose the first project the filesystem gives us": More complicated apps will have multiple projects (libraries, app, support modules, etc.). Is this supposed to find the one that's the app? can we sniff which of the vcxprojs are type = Application? (is that the right thing to sniff?)

jonthysell replied:
We can technically make the heuristic as complicated as we like, but this covers the basic case: one solution, one project, named according to what the CLI produces. This backup is in-case a different name was chosen, but are still one solution, one project.

For modules/apps where this heuristic fails, they can always define the correct values in their own react-native.config file.

asklar replied:
the problem is that this sets up a tar pit that will take running into an issue and trying to diagnose what went wrong, whereas we know this is a problem even for an in-repo app and we can set up guardrails to protect people from setups that we can't figure out.

jonthysell replied:
Unfortunately I can't add command-line arguments into react-native config, we're just plugging into their cli. So I can make it fail if there's more than one project and we don't know which to choose, but the correct fix is for people to use react-native.config and tell us.

jonthysell replied:
Fixed: if the single correct file (solution or project) can't be idnetified, an error will be thrown and the developer will have to use the react-native.config.

acoates-ms commented on vnext/local-cli/config/dependencyConfig.js:
```diff
@@ -1,102 +1,117 @@
+      projectFile: '', // relative path to the project, ex: ProjectName\ProjectName.vcxproj
+      projectName: '', // name of the project
+      projectLang: '', // language of the project, cpp or cs
+      projectGuid: '', // project identifier
```
Is this any guid that the package wants to generate, or does it have to match a guid used by ProjectName.vcxproj?  And if it has to match, can we just extract it from the project file?

jonthysell replied:
We do extract it, but the process is decoupled. react-native config does all the extracting and then serves up the data for autolinking (and others?). But I guess for projects where they create their own react-native.config file, we could add some logic to extract it again in case they forget it.

jonthysell replied:
Fixed.

asklar commented on vnext/local-cli/generator-windows/templates/cpp/src/AutolinkedNativeModules.g.h:
```diff
@@ -1,17 +1,14 @@
 namespace winrt::Microsoft::ReactNative
 {
 
 static void RegisterAutolinkedNativeModulePackages(winrt::Windows::Foundation::Collections::IVector<winrt::Microsoft::ReactNative::IReactPackageProvider> const& packageProviders)
```
I don't think this is quite right - static in the context of a free function (even in a namespace) means it will only be visible to the translation unit it's in (i.e. it's going to be visible only to the CPP that includes this header).
in this case it's no-op because there's no code, but when you start adding code to it, you'll end up with duplicate copies of  the code. Note that you won't get duplicate definitions because its visibility is limited to the file that includes the header.

jonthysell replied:
It's only called once, in the template during app init - this originally was a macro just to insert the code into that app init location, but I was asked to make it into a function. No-one complained when I put it into the template weeks ago, and changing will require changing the template. Is there a better option?

asklar replied:
yeah I had noticed in the PR but forgot to comment, my bad :)
there are a few options. The right thing would be to have the declaration of the function in this header file:
```cpp
void RegisterAutolinkedNativeModulePackages(winrt::Windows::Foundation::Collections::IVector<winrt::Microsoft::ReactNative::IReactPackageProvider> const& packageProviders);
```
Then have its definition in an associated cpp file.

2nd option is to put this function into a class as a static method.

3rd less nice option is to mark the function as `__declspec(selectany)`

jonthysell replied:
This is what the cli will make for 62 apps - if I change it here then we should backport it into 0.62.

jonthysell replied:
I have a separate PR out for this fix here: https://github.com/microsoft/react-native-windows/pull/5090

That way I can get just this change backported into 62 so that apps made in 62 are set up for autolinking in 63.

jonthysell replied:
fixed.

asklar commented on vnext/local-cli/runWindows/runWindows.js:
```diff
@@ -51,8 +52,9 @@ async function runWindows(config, args, options) {
 
   if (options.autolink) {
-    autolink.updateAutoLink(verbose);
+    await autolink.func(null, null, {logging: options.logging});
```
names for the params? if we pass an options param, maybe make the 1st two params be part of that dictionary so you can name what those nulls are more easily

jonthysell replied:
The structure for the command is the format that react-native subcommands use. I suppose I could pass in the values that were passed into `run-windows`. I don't use them within auto-linking.

jonthysell replied:
Fixed.

asklar commented on vnext/local-cli/runWindows/utils/vstools.js:
```diff
@@ -0,0 +1,176 @@
+const path = require('path');
+const chalk = require('chalk');
+
+function hasBlock(lines, block) {
```
name this more explicitly? what kind of block are we talking about?

jonthysell replied:
Fixed.

asklar commented on vnext/local-cli/runWindows/utils/vstools.js:
```diff
@@ -0,0 +1,176 @@
+const chalk = require('chalk');
+
+function hasBlock(lines, block) {
+  var startIndex = lines.indexOf(block[0]);
```
check for `block.length > 0` ?

jonthysell replied:
Fixed.

asklar commented on vnext/local-cli/runWindows/utils/vstools.js:
```diff
@@ -0,0 +1,176 @@
+
+  const projectTypeGuid =
+    project.projectLang === 'cpp'
+      ? '{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}'
```
create const value to name what these guids represent

jonthysell replied:
Done.

asklar commented on vnext/local-cli/runWindows/utils/vstools.js:
```diff
@@ -0,0 +1,176 @@
+  slnConfigs.forEach(slnConfig => {
+    projectConfigLines.push(
+      `\t\t${projectGuid}.${slnConfig}.ActiveCfg = ${slnConfig.replace(
+        'x86',
```
we might want to be more cautious here (and below), e.g. if my project is named somethingx86 then the wrong thing will get replaced.
Or add a block in the project creation that the name cannot contain x86

jonthysell replied:
This line only executes on config names (x86, x64, arm, arm64)  not project names.

asklar commented on vnext/local-cli/runWindows/utils/vstools.js:
```diff
@@ -0,0 +1,176 @@
+  projectConfigLines.forEach(projectConfigLine => {
+    if (slnLines.indexOf(projectConfigLine) < 0) {
+      if (verbose) {
+        console.log(chalk.yellow('Missing project config block.'));
```
nit if think there's a newWarn function that will output this in the progress thing?

jonthysell replied:
Yeah but that function writes every line twice to get the functionality it supports. It's not really a warning, just info., I only colored it to be a little more visible.

acoates-ms commented on vnext/local-cli/config/dependencyConfig.js:
```diff
@@ -1,102 +1,117 @@
+      projectGuid: '', // project identifier
+    },
+  ],
+  packages: [ // array of nuget packages that include modules
```
This should be called nugetPackages or something.  RN developers are going to think NPM packages when they just see packages.

And this whole thing is just needed because VS C++ projects can't include nuget references directly right?  I'm worried that we are kinda trying to implement a VS feature here..  Shouldn't we just get VS to fix C++ nuget support?  And if they dont.. then I'm not sure we should take that on.   Also why does this section need another set of cppPackageProviders etc?  Can't they just be specified by one of the projects in the first list?  (projects.cppPackageProviders)

jonthysell replied:
No, this is to support the scenario where a native module publishes to npm and chooses not to publish windows source in their npm package, but instead publishes a nuget binary. The react-native config can't possibly figure that out on it's own, so they'll have to create a react-native.config file that specifies the nuget package name and version string that matches npm package.

Autolinking will then read this info and use it to either add packagereferences in the generated targets file (for C#) or modify the app's packages.config (for C++).

I can change the name to nuget package to make it more clear.

jonthysell replied:
It at some point VS does support packagereferences for C++, then we just update auto-linking to add packagereferences for both C# and C++ and stop touching packages.config.

jonthysell replied:
Fixed.

acoates-ms commented on vnext/local-cli/config/dependencyConfig.js:
```diff
@@ -1,102 +1,117 @@
+      csPackageProviders: [], // array of fully qualified cs IReactPackageProviders, ie: 'ModuleName.ReactPackageProvider'
+    },
+  ],
+  additionalProjects: [ // array of (dependency) projects that must be built, but do not contain modules
```
Can we simplify this and just allow projects to not have any package providers, then you just need the one list of projects?

jonthysell replied:
We need a way of differentiating:
* These projects need to be both in the solution and directly referenced by the app project (then we loop through the package providers, if any, to be added to the project)
* These projects need to be in the solution just so the build works. I could refactor those into a boolean

jonthysell replied:
Fixed.

asklar commented on vnext/local-cli/runWindows/utils/autolink.js:
```diff
@@ -6,95 +6,359 @@
+namespace winrt::Microsoft::ReactNative
+{
+
+static void RegisterAutolinkedNativeModulePackages(winrt::Windows::Foundation::Collections::IVector<winrt::Microsoft::ReactNative::IReactPackageProvider> const& packageProviders)
```
same comment about `static` here

jonthysell replied:
Fixed.

asklar commented on vnext/local-cli/runWindows/utils/autolink.js:
```diff
@@ -6,95 +6,359 @@
+
+      const cppHeaderContents = `// AutolinkedNativeModules.g.h contents generated by "react-native autolink-windows"
+
+#include "pch.h"${cppIncludes}
```
consider reusing the logic we have in the project creation  code to do variable substitution, and have this content be in separate files instead of being strings in code

NickGerleman replied:
+1 to this. This could help avoid the template getting out of sync.

jonthysell replied:
Done.

asklar commented on vnext/local-cli/runWindows/utils/autolink.js:
```diff
@@ -6,95 +6,359 @@
+    },
+    {
+      command: '--check',
+      description: 'Check (only) if any autolinked files need to change',
```
consider writing this as "Only check whether any autolinked files need to change", or "Check if any autolinked files need to change and exit"

jonthysell replied:
Done.

NickGerleman commented on vnext/local-cli/config/configUtils.js:
```diff
@@ -0,0 +1,144 @@
+      '**/Debug/**',
+      '**/Release/**',
+      '**/Generated Files/**',
+      '**/packages/**',
```
Blacklisting any directory or subdirectory named "packages" feels a bit broad. Could we remove or tighten this one?

jonthysell replied:
Unfortunately that's where nuget packages get restored and we don't want to search there. We could remove the path, but I'd rather leave it here and if someone happens to hit the small chance to have a conflict, they can use the override.

NickGerleman replied:
I mean that we do this right now for anything named packages. E.g `foo/bar/packages`, instead of just the one we know we care about.

NickGerleman replied:
We've hit this all the time even in our repo with people being to general with .gitignore files for example. If we know where a directory lives, we shouldn't blacklist every subdirectory that happens to have the same name.

jonthysell replied:
The glob is run on the working directory which is the windows folder under the particular app/ native module lib root. There are no packages folders under windows folders in our repo, nor do we put ones there for new apps. The *only* case this is a problem is if a user literally creates an RN app named "packages" and then runs our init, creating a windows folder with `packages.sln` and a `packages\packages.vcxproj`, in which case they're probably going to have a lot of problems with VS anyway. Won't fix.

NickGerleman commented on vnext/local-cli/config/configUtils.js:
```diff
@@ -0,0 +1,144 @@
+
+const xmldoc = require('xmldoc');
+
+// find VS files matching the pattern
```
Nit: JS tends to use JSDoc with has Javadoc style "/**". Vs code will auto close it for you.

jonthysell replied:
Fixed.

NickGerleman commented on vnext/local-cli/config/configUtils.js:
```diff
@@ -0,0 +1,144 @@
+  } else if (projectPath.endsWith('.csproj')) {
+    return 'cs';
+  }
+  return null;
```
Do we try to handle null values in callers? Consider throwing if something unexpected happens instead.

jonthysell replied:
Fixed.

NickGerleman commented on vnext/local-cli/runWindows/utils/autolink.js:
```diff
@@ -6,95 +6,359 @@
+const path = require('path');
+const chalk = require('chalk');
+
+const {newSpinner} = require('./commandWithProgress');
```
If we're using a spinner or trying to be interactive we should try to avoid the synchronous fs calls we added and await fa.promises, etc instead. We can otherwise get something that stutters.

Looks like some code is already async?

jonthysell replied:
I can change this in a future iteration.

NickGerleman commented on vnext/local-cli/runWindows/utils/autolink.js:
```diff
@@ -6,95 +6,359 @@
+  return contentsChanged;
+}
+
+function ExitProcessWithStatusCode(statusCode, loggingWasEnabled) {
```
Nit: lowercase

jonthysell replied:
Done.

NickGerleman commented on vnext/local-cli/runWindows/utils/autolink.js:
```diff
@@ -6,95 +6,359 @@
-      const packageRegistration = `    \\\r\n    PackageProviders().Append(winrt::${trimmedPackageName}::ReactPackageProvider());`;
-      packageRegistrations += packageRegistration;
+
+      const cppHeaderContents = `// AutolinkedNativeModules.g.h contents generated by "react-native autolink-windows"
```
Do we want to disable Clang format for this as well?

jonthysell replied:
We need to disable the clang formatting in the template files so our repo formatting settings don't touch it. But the CLI actually removes that line from the file when it creates a new app, because it's not for us to decide if an app repo wants to use clang, or not, and what formatting settings they use.

NickGerleman commented on vnext/local-cli/runWindows/utils/autolink.js:
```diff
@@ -6,95 +6,359 @@
+    ExitProcessWithStatusCode(1, verbose);
   }
+
+  return;
```
Nit: can this be removed?

jonthysell replied:
Done.

NickGerleman commented:
I mentioned this somewhere else, but we should really consider moving local CLI code to typescript with implicit any sooner than later. Types can be incrementally added, and we get way better readability, dev experience, the ability to use ES6 import and export, etc.

NickGerleman commented:
There has been a lot of recent conversation about how we need better test coverage. Any ideas on how we might be able to build validation for this change?

jonthysell commented on change/react-native-windows-2020-05-28-13-42-40-autolink.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "prerelease",
+  "comment": "Auto-linking for native modules (source-based) * Updated `npx react-native config` to create the needed json for auto-linking * New `npx react-native autolink-windows` command to perform auto-linking * Auto-linking occurs automatically during `npx react-native run-windows` * New Auto-linking targets and props so CLI apps throw a warning if auto-linking should be run * New targets an ...
```
Fixed.

NickGerleman commented:
> > There has been a lot of recent conversation about how we need better test coverage. Any ideas on how we might be able to build validation for this change?
> 
> The current CI will already capture the case of a CLI app where there are no community modules installed. Autolinking will run, and everything should still work.
> 
> At some point we'll want a CI task that creates a new app, adds a community module, builds and verifies that autolinking worked. However to do that we need to get this in, and backported into a release, and get a community module that is willing/able to upgrade to the RN version we've backported to, then make sure the windows implementation is upgraded and working too.
> 
> In that respect I'm working on a PR for react-native-camera, getting it to build against 0.62, but they are built on RN 0.59, and I'm not sure yet how to make the windows implementation build on both 0.62 and anything earlier, since 0.62 changed a lot build-wise.

Using an external community module in our CI would cause breakage any time we move our repo forward.  We're set up as a monorepo with multiple npm packages already. Can't we have a test community module in the repo if we want something end-to-end? It feels like we can do better than trying to validate by backporting and changing a real module. Autolinking in 0.62 after release would be a worthwhile triage discussion to have.

Apart from something end-to-end, do we have individual bits of testable logic? It's really hard to motivate ourselves to add test coverage after the fact.

jonthysell commented:
> Using an external community module in our CI would cause breakage any time we move our repo forward. We're set up as a monorepo with multiple npm packages already. Can't we have a test community module in the repo if we want something end-to-end? It feels like we can do better than trying to validate by backporting and changing a real module. Autolinking in 0.62 after release would be a worthwhile triage discussion to have.
> 
> Apart from something end-to-end, do we have individual bits of testable logic? It's really hard to motivate ourselves to add test coverage after the fact.

Since #4987 propagated the use of the new app target files into all of the apps in our repo, which is where this autolinking work happens, I've now had to enable autolinking for all of those projects. So autolinking will run for both SampleApps, playgrounds, and the E2ETest, even though none of those rely on any community modules (so still justing that case).

While I like the idea of having a "test" community module in this repo, I'm not sure how to make it also consumable by a test app in this repo. Either I'd have to so some yarn workspaces stuff, or maybe I can extend the CI to both publish rnw and the community module to verdaccio, then create a new app, add the module and confirm it works.

jonthysell commented on vnext/PropertySheets/Autolink.targets:
```diff
@@ -0,0 +1,10 @@
+ Licensed under the MIT License.. 
+-->
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Target Name="RunAutolink" BeforeTargets="PrepareForBuild">
```
What expected inputs? The commands and args are defined in the props file, it still works if there are no community modules found.

asklar replied:
> What expected inputs? The commands and args are defined in the props file, it still works if there are no community modules found.

@jonthysell in order for MSBuild to know that it's safe to not re-run a Target, it has to know what the inputs and outputs are. If the existing outputs are older than the inputs' last modified timestamp (and command line options haven't changed), then msbuild will not re-run a target. You'd specify the list of files that are inputs and outputs as an attribute to the Target.
See: https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-targets?view=vs-2019

jonthysell replied:
There are no input files, the only "input" is the result of the `react-native config` command (which is run within the autolinkin command), which I don't want to write to disk. And since this just does a check (with no files updated) there are no output file either.

NickGerleman replied:
FYI @vmoroz since there was some recent discussion about our incremental builds, and cutting down on anything that we can't do incrementally. I'm okay with this for now personally though, if it's hard to make this incremental. @jonthysell do you know current built-time impact?

An important bit that I forgot to brought up is that Visual Studio normally does a dumb thing with build logic where it will skip running targets entirely if filter files don't change. I disable this right now in `ReactPatches.targets` by doing this:

```xml
  <PropertyGroup>
    <DisableFastUpToDateCheck>True</DisableFastUpToDateCheck>
  </PropertyGroup>
```

Removing that would break this as well, so we should consider putting that in a place more global if we are going to continue adding targets not derived from filter files.

jonthysell replied:
The autolinking check only occurs on app project builds and runs in ~500-600 milliseconds. ReactPatches.targets is only imported for our libraries, not app projects. Changes there won't have any impact on the app projects.

NickGerleman replied:
`ReactPatches.targets` is still run as part of the MSRN build. The prop is weird and applies to the current Visual Studio instance instead of a specific project, so it currently leaks pretty far. You can test by removing the prop then restarting Visual Studio, but it definitely has an impact on user projects right now.

jonthysell replied:
The worse case scenario if this target doesn't run - if the dev changes their depdencies on community modules (add, remove, update) and hasn't run re-run auto-linking once (typically by running run-windows), they won't be warned that they should run the command. If their build breaks because that last time auto-linking, the generated files are now invalid, then they'll see an error in those generated files, and see the command they should run.

NickGerleman replied:
I don't mean to keep harping on this, but we have concrete user scenarios where the wrong thing will happen. Could we have an conversation on any concerns preventing adding this single prop to something reachable from external projects? I understand there's workarounds for when this fails, but unless I'm missing something, the fix here should be pretty trivial.

There's been consensus from multiple engineers that IDE support is something that we do care about. Without DisableFastUpToDateCheck defined somewhere we won't run the target unless native build files change. This works now because the property affects the global state of Visual Studio, so it affecting internal projects also impacts MSRN projects.

This won't be the case when we get to binary distribution, so we'll suddenly see autolinking break for the IDE scenario.

jonthysell replied:
Auto linking changes project files. It cannot happen during the build without causing problems.

This target only checks if the files could be updated by autolinking, and warns the dev to run auto-linking.

If someone builds in VS, and _for any reason_, this check doesn't run, the *only* breaking scenario is a build break because a community module was uninstalled and auto-linking not re-run, because the previously generated files reference that missing module. The build break will be in the files that tells user what to run to fix it.

In that scenario there are _no filter files_ that changed that we could have checked for in the first place.

If you have an explicit change that fixes this, submit it. Otherwise, there's nothing to be done.

asklar commented on packages/E2ETest/windows/ReactUWPTestApp/ReactUWPTestApp.csproj:
```diff
@@ -1,10 +1,9 @@
+  <PropertyGroup Label="ReactNativeWindowsProps">
     <ReactNativeWindowsDir Condition="'$(ReactNativeWindowsDir)' == ''">$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), 'node_modules\react-native-windows\package.json'))\node_modules\react-native-windows\</ReactNativeWindowsDir>
   </PropertyGroup>
-  <Import Project="$(ReactNativeWindowsDir)\PropertySheets\external\Microsoft.ReactNative.Uwp.CSharpApp.props"/>
```
please sync with @dannyvv regarding usage props/targets :)

jonthysell replied:
I moved this to be consistent with the CLI template.

asklar commented on packages/microsoft-reactnative-sampleapps/windows/SampleAppCPP/SampleAppCpp.vcxproj:
```diff
@@ -193,9 +195,6 @@
     </Text>
   </ItemGroup>
-  <ItemGroup>
-    <Natvis Include="$(ReactNativeWindowsDir)\Folly\Folly.natvis" />
```
why are we removing the natvis ?

jonthysell replied:
App shouldn't need to reference this since they're consuming Microsoft.ReactNative, not Folly directly. Right?

jonthysell replied:
IF so, this should be added to the shared targets for everyone to enjoy, not just this one app.

asklar replied:
I believe this file is used by VS to format Folly objects in the debugger, it's not used for building the app or at runtime

NickGerleman replied:
Users of MSRN shouldn't directly touch Folly, so I can understand this being removed, but the natvis is useful for engineers on RNW debugging internals. Don't think it's needed in something user-facing though.

asklar commented on packages/microsoft-reactnative-sampleapps/windows/SampleAppCS/App.xaml.cs:
```diff
@@ -42,7 +42,9 @@ public App()
             InstanceSettings.Properties.Set(ReactPropertyBagHelper.GetName(null, "Prop2"), "Hello RNW!");
 
-            PackageProviders.Add(new ReactPackageProvider()); // Includes any modules in this project
+            Microsoft.ReactNative.Managed.AutolinkedNativeModules.RegisterAutolinkedNativeModulePackages(PackageProviders); // Includes any autolinked modules
```
is this call ok to make before we modify PackageProviders? should it not happen after we add everything?

asklar commented on packages/playground/react-native.config.js:
```diff
@@ -0,0 +1,12 @@
+module.exports = {
+  // Uncomment the below when running autolink-windows for playground-win32
```
when would that be the case? e.g. is autolinking something people need to do when they do XYZ? (what is XYZ?)

jonthysell replied:
Fixed.

asklar commented on vnext/PropertySheets/External/Microsoft.ReactNative.Uwp.CSharpApp.targets:
```diff
@@ -18,11 +22,12 @@
-  <Import Project="$(ReactNativeWindowsDir)PropertySheets\Bundle.props" />
-  <Import Project="$(ReactNativeWindowsDir)PropertySheets\Bundle.targets" />
+   <!-- The props file for bundling is not set up to be just defaults, it assumes to be run at the end of the project. -->
+  <Import Project="$(ReactNativeWindowsDir)\PropertySheets\Bundle.props" />
```
FYI directories variables in MSBuild often already include the trailing backslash (as is the case here). Adding the \ means it will look at foobar\\barbaz which is innocuous but unnecessary

jonthysell replied:
Unfortunately `ReactNativeWindowsDir` is a variable we explicitly set and expose in end user app project files, to give them the chance to override them (as we do within our in-repo projects to handle the mono-repo). Adding the `\` is an easy protection in case they forget.

asklar commented on vnext/PropertySheets/External/Microsoft.ReactNative.Uwp.Common.props:
```diff
@@ -0,0 +1,12 @@
+  Do not make any changes here unless it applies to ALL such projects.
+-->
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$(MSBuildThisFileDirectory)Microsoft.ReactNative.Common.props" />
```
see for ex here, we don't use a trailing backslash after MSBuildThisFileDirectory

jonthysell replied:
That is a essentially a read-only value that is always set correctly.

asklar commented on vnext/local-cli/config/dependencyConfig.js:
```diff
@@ -1,102 +1,167 @@
+      'csPackageProviders',
+    ];
+
+    for (let i = 0; i < projects.length; i++) {
```
nit, `for... of` ?

jonthysell replied:
Fixed.

asklar commented on vnext/local-cli/config/dependencyConfig.js:
```diff
@@ -1,102 +1,167 @@
-    ignore: ['node_modules/**', '**/Debug/**', '**/Release/**', '**/Generated Files/**', '**/packages/**'],
-  })[0];
+    const cppNamespace = projectNamespace.replace('.', '::');
+    const csNamespace = projectNamespace.replace('::', '.');
```
careful with this, I was bitten by this last week - make sure you replace all instances since the default is to do just the first. Instead, do `.replace(/\./g, '::')` and `.replace(/::/g, '.')`

asklar commented on vnext/local-cli/runWindows/utils/autolink.js:
```diff
@@ -6,95 +6,341 @@
+      for (const dependencyName in windowsDependencies) {
+        windowsDependencies[dependencyName].projects.forEach(project => {
+          if (project.directDependency) {
+            csUsingNamespaces += `\n\n// Namespaces from ${dependencyName}`;
+            project.csNamespaces.forEach(namespace => {
+              csUsingNamespaces += `\nusing ${namespace};`;
+            });
+
+            csReactPacakgeProviders += `\n            // IReactPackageProviders from ${dependencyName}`;
+            project.csPackageProviders.forEach(packageProvider => {
+              csReactPacakgeProviders += `\n            packageProviders.Add(new ${packageProvider}());`;
+            });
+          }
+        });
+      }
```
i may have said this elsewhere but please consider making this a templated file like how we do in the app template code :)

asklar commented on vnext/local-cli/runWindows/utils/autolink.js:
```diff
@@ -6,95 +6,341 @@
-      const packageRegistration = `    \\\r\n    PackageProviders().Append(winrt::${trimmedPackageName}::ReactPackageProvider());`;
-      packageRegistrations += packageRegistration;
+
+      const cppContents = `// AutolinkedNativeModules.g.cpp contents generated by "react-native autolink-windows"
+#include "pch.h"
+#include "AutolinkedNativeModules.g.h"${cppIncludes}
+
+// clang-format off
+namespace winrt::Microsoft::ReactNative
+{
+
+void RegisterAutolinkedNativeModulePackages(winrt::Windows::Foundation::Collections::IVector<winrt::Microsoft::ReactNative::IReactPackageProvider> const& packageProviders)
+{${cppPackageProviders}
+}
+
+}
```
same comment as C# of using a template file

jonthysell replied:
Done.

jonthysell commented:
Added Breaking Change since the template is being touched (even though it's not really breaking).

dannyvv commented on packages/microsoft-reactnative-sampleapps/windows/SampleAppCS/AutolinkedNativeModules.g.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Microsoft.ReactNative.Managed
+{
+    internal static class AutolinkedNativeModules
```
For C# it could be done a bit cleaner by having a regular file with a partial method checked in and generate into the OBJ tree, and same with the .targets file for the referneces, to have that in the obj directory.

But I recognize that the convention for react native developers is to check in generated files... I still get a jolt that this is a big no-no. 

Maybe beceause the file is a `.g.cs`. Where it is more a `.lock.cs` since it only gets updated when you run the autolink tool. to me a .g.cs means it is regeenrated on every boyd...

jonthysell replied:
It will get re-generated on every build, that's how autolinking works.

I'm not going to change this any further, every request to change this means yet another separate breaking change PR I have to backport into 62.

NickGerleman commented:
> While I like the idea of having a "test" community module in this repo, I'm not sure how to make it also consumable by a test app in this repo. Either I'd have to so some yarn workspaces stuff, or maybe I can extend the CI to both publish rnw and the community module to verdaccio, then create a new app, add the module and confirm it works.

The Yarn workspaces stuff should happen pretty automatically. That package will get symlinked in node_modules, so any updates should be reflected. This is how Playground, etc automatically pick up changes we make to react-native-windows.

Having a sample community module as a template could be nice for other things. Like, I've seen 0.62 community modules use outdated metro configs, so centralizing "This is the thing to copy" could be nice.

jonthysell commented:
> Requesting changes around reliability and scope of IDE integration. There was some discussed in resolved comments for fixes for known reliability issues, but the response also seemed to indicate we weren't ever going to run autolinking from the IDE and warn instead to manually link. I missed this in the description personally, and that sounds like a reasonable tradeoff, but it's becoming clearer to me that we're not on the same page as a team on where this is headed and what we're supporting.
> 
> In general there have been enough interest and concerns raised around autolinking in comments here and elsewhere that I think it would be beneficial to have some sort of design discussion or doc. It would be great to hear all of the context and plans you've been working on 🙂. Don't think we have any need to be perfect to start iterating by any means, but I we should also be careful not to immediately dismiss the concerns that have been raised.

@NickGerleman, please open a separate issue concerning the Targets affected by ` <DisableFastUpToDateCheck>True</DisableFastUpToDateCheck>` as there are others besides this one. The solution may be to update those targets, but it may be to add ` <DisableFastUpToDateCheck>True</DisableFastUpToDateCheck>` into the app targets that will still get loaded. It's out of scope for this PR.

I've addressed all feedback provided.

asklar commented on vnext/local-cli/config/configUtils.js:
```diff
@@ -0,0 +1,259 @@
+    cwd: folder,
+    ignore: [
+      'node_modules/**',
+      '**/Debug/**',
```
nit, should this be also ignoring WinUI3 since that's another config?

jonthysell replied:
Sure.

asklar commented on vnext/local-cli/config/configUtils.js:
```diff
@@ -0,0 +1,259 @@
+}
+
+/**
+ * Checks if the target file path is a RNW solution file.
```
```suggestion
 * Checks if the target file path is a RNW solution file by checking if it contains the string ReactNative
```

jonthysell replied:
Fixed.

asklar commented on vnext/local-cli/config/configUtils.js:
```diff
@@ -0,0 +1,259 @@
+    fs
+      .readFileSync(filePath)
+      .toString()
+      .search(/Microsoft\.ReactNative\.Uwp\.(Cpp|CSharp)App\.targets/) > 0
```
this seems fragile.
Can we instead add a property in the project template that just tells us which one is the app project? something like
```xml
<PropertyGroup>
    <ReactNativeWindowsProjectType>App</ReactNativeWindowsProjectType>
</PropertyGroup>
```
then just read the value of this property?

jonthysell replied:
There are already two user-controlled mechanisms for specifying the right project if one can't be determined - the official `react-native.config.js` file approach and our `--proj` args to `run-windows` and `autolink-windows`. Adding another one seems overkill and confusing, especially if it means modifying the template. Plus autolinking won't work unless you're including that app target file anyway (because that target is what makes sure the generated target gets included).

If you're the .01% who is creating multiple RNW app projects in your code, or removing the import target that is necessary for making sure a RNW app builds, then I think you're responsible for also using any of the two officially supported overrides.

asklar commented on vnext/local-cli/config/configUtils.js:
```diff
@@ -0,0 +1,259 @@
+ * @param {string} tagName The XML tag to look for.
+ * @return {string} The value of the tag if it exists.
+ */
+function findTagValue(projectContents, tagName) {
```
can we just use JavaScript's XML parsing instead? this would break if we ever added even a space or a condition.
For example:
```xml
<MyTag >value</ MyTag>
```
or
```xml
<MyTag Condition="foo">value</MyTag>
```
would fail. Using JavaScript's XML parsing would let us at least deal with the first case, although the 2nd case would need more work/integration with MSBuild that's probably hard to make happen.

jonthysell replied:
Is there one that we're already using? I don't want to add a new xml parser dependency to our repo. There's like a million XML parsers, and the one we were including as a dependency before was junk.

jonthysell replied:
Switching to using `xmldom` and `xpath`, thouugh that means adding a dependency.

asklar commented on vnext/local-cli/config/dependencyConfig.js:
```diff
@@ -1,102 +1,244 @@
+    // Manaully provided projects, so extract the sourceDir
+    if (!('sourceDir' in userConfig)) {
+      sourceDir =
+        'Error: Source dir is required if projects are specified, but it is not specified in react-native.config.';
```
throw this string instead of returning it? (here and below)

jonthysell replied:
I started with that, but unfortunately, the structure of the config command is that it disables logging and returns a JSON blob output or an error. Having a single error thrown breaks the whole command, which means you also break the entire CLI and it's commands, even for other platforms. That's a crappy side effect, and it doesn't even catch when there are multiple errors (say if you have errors in your app config and there's errors in some dependency's config).

The only other option is to fail silently by leaving filed null, then having the caller of `react-native config` go figure out why it was wrong, or set the field to an error that we can see and process later in our consumption of the config.

asklar commented on vnext/local-cli/config/dependencyConfig.js:
```diff
@@ -1,102 +1,244 @@
+
+      const directDependency = true;
+
+      const cppNamespace = projectNamespace.replace(/\./g, '::');
```
I _think_ @dannyvv  made it so the project namespace always uses dots so this could be simplified if you wanted

jonthysell replied:
I just wanted to cover my bases.

asklar commented on vnext/local-cli/config/projectConfig.js:
```diff
@@ -1,66 +1,164 @@
-  })[0];
+    // Manual override, try to use it for project
+    if (!('project' in userConfig)) {
+      result.project =
```
throw? here and below

jonthysell replied:
As above.

asklar commented on vnext/local-cli/config/projectConfig.js:
```diff
@@ -1,66 +1,164 @@
+    } else if (foundProjects.length > 1) {
+      result.project = {
+        projectFile:
+          'Error: Too many app project files found, please specify in react-native.config.',
```
Do I understand correctly that this requires editing the config file to hint to the CLI which solution to build?
Users should be able to build any project with `npx react-native run-windows` with some arguments if it isn't obvious which solution/project to build. Storing these things in the config adds friction. For example, if we wanted to build both playground and playgroun-win32, do we now need to edit the config file in the pipeline ? is this config file checked in?

asklar replied:
It looks like this still uses the `--sln`  option so it sounds like there are more than one way which isn't bad but we just need to be careful with documenting this. Can you include a markdown doc pretty please :) ?
The way I understand it the behavior is:
- if user passes `--sln` then use that solution
- otherwise, check `react-native.config` for which solution to use

Then to determine the app project we use the existence of the right `.targets` file `Import` statement which I'd like to make more explicit and instead use a particular property in the project so that it's clear what this does and we aren't tied to the name or existence of the targets import

jonthysell replied:
`react-native.config` is the official way to override this for the cli. It's at the core of the how the cli operates, all cli commands are supposed to use the information provided by config. This is why I was saying that our RNW CLI commands are fundamentally not integrating with RN commands, duplicating functionality and making it more confusing for RN developers who expect to modify the config if necessary. I think it's the wrong approach to optimize for the 0.01% case that playground represents.

asklar commented on vnext/local-cli/generator-windows/templates/cpp/src/AutolinkedNativeModules.g.targets:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <!-- AutolinkedNativeModules.g.targets contents generated by "react-native autolink-windows" -->
+  <ItemGroup>{{ &autolinkProjectReferencesForTargets }}
```
If I read the doc correctly it sounds like & is used to render unescaped HTML, meaning that by default it escapes things as if they were to end up in HTML. That's a weird default for mustache...

Should we standardize on how we use mustache for unescaped text? it looks like you can also use {{{ }}} (triple braces) which might be slightly less eye-sore-y :)

jonthysell replied:
Mustache was for templating HTML, which is why it escapes by default. Since we're using it to template code, especially code in a language that uses curly braces, adding another set of curly braces makes it harder to spot. I'd rather have an explicit thing you can see (there are other characters used by mustache for other things like ^ conditionals).

Since the decision was made to use mustache everyone needs to learn mustache.

jonthysell commented:
I can't check-in the docs until the 0.62 releases and the website version is bumped.

Here's the PR: https://github.com/microsoft/react-native-windows-samples/pull/145

Here are the relevant docs:
* https://github.com/jonthysell/react-native-windows-samples/blob/autolink/docs/config.md
* https://github.com/jonthysell/react-native-windows-samples/blob/autolink/docs/native-modules-autolinking.md

asklar commented on vnext/PropertySheets/Autolink.props:
```diff
@@ -8,8 +8,9 @@
     <AutolinkCommand Condition="'$(AutolinkCommand)' == ''">npx react-native autolink-windows</AutolinkCommand>
-    <AutolinkCommandArgs Condition="'$(AutolinkCommandArgs)' == ''">--check</AutolinkCommandArgs>
     <AutolinkCommandWorkingDir Condition="'$(AutolinkCommandWorkingDir)' == ''">$([MSBuild]::GetDirectoryNameOfFileAbove($(ProjectDir), 'package.json'))</AutolinkCommandWorkingDir>
+    <AutolinkCommandArgs Condition="'$(AutolinkCommandArgs)' == '' And '$(SolutionPath)' != '' And '$(ProjectPath)' != ''">--check --sln $([MSBuild]::MakeRelative($(AutolinkCommandWorkingDir), $(SolutionPath))) --proj $([MSBuild]::MakeRelative($(AutolinkCommandWorkingDir), $(ProjectPath)))</AutolinkCommandArgs>
```
who's setting SolutionPath and ProjectPath?

jonthysell replied:
I added this as an extra safeguard to make sure the check runs accurately for playground (and potential other multi-solution or multi-app project setups). It's conditional on SolutionPath and ProjectPath existing. Otherwise, it falls through to the default args which is to just use the result from `react-native config`.

Alternatively I could just override the args prop in playground projects, but I didn't want to make such a change in the template for all apps.

I know running msbuild directly against a project doesn't set the solution path. But  using VS or `run-windows`, it gets set.

jonthysell replied:
I'm not 100% happy with it, as we *should* be relying entirely on `react-native config`, but this was the best compromise I could come up with for multi solution setups.

asklar replied:
@jonthysell  thanks for the explanation. I can't find how it gets set through run-windows though (I assume VS has its own way of invoking msbuild which sets these variables up), but since we're invoking msbuild manually I would have expected us to pass /p:... or something like that
